compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeWindowDelegate.desktop.kt

```diff
@@ -60,30 +60,123 @@ internal class ComposeWindowDelegate(
+                }
+                field = value
+                bridge.isWindowTransparent = value
+                bridge.transparency = value || interopBlending
+
+                /*
+                 * Windows makes clicks on transparent pixels fall through, but it doesn't work
+                 * with GPU accelerated rendering since this check requires having access to pixels from CPU.
+                 *
+                 * JVM doesn't allow override this behaviour with low-level windows methods, so hack this in this way.
+                 * Based on tests, it doesn't affect resulting pixel color.
+                 *
+                 * Note: Do not set isOpaque = false for this container
+                 */
+                if (value && hostOs == OS.Windows) {
+                    pane.background = Color(0, 0, 0, 1)
+                    pane.isOpaque = true
+                } else {
+                    pane.background = null
+                    pane.isOpaque = false
+                }
+
+                window.background = if (value && !skikoTransparentWindowHack) Color(0, 0, 0, 0) else null
+            }
+        }
+
+    /**
+     * There is a hack inside skiko OpenGL and Software redrawers for Windows that makes current
+     * window transparent without setting `background` to JDK's window. It's done by getting native
+     * component parent and calling `DwmEnableBlurBehindWindow`.
+     *
+     * FIXME: Make OpenGL work inside transparent window (background == Color(0, 0, 0, 0)) without this hack.
+     *
+     * See `enableTransparentWindow` (skiko/src/awtMain/cpp/windows/window_util.cc)
+     */
+    private val skikoTransparentWindowHack: Boolean
+        get() = hostOs == OS.Windows && renderApi != GraphicsApi.DIRECT3D
+
     private val _pane = object : JLayeredPane() {
         override fun setBounds(x: Int, y: Int, width: Int, height: Int) {
-            bridge.component.setSize(width, height)
+            bridge.component.setBounds(0, 0, width, height)
             super.setBounds(x, y, width, height)
         }
 
         override fun add(component: Component): Component {
-            val clipComponent = ClipComponent(component)
-            clipMap[component] = clipComponent
-            bridge.clipComponents.add(clipComponent)
-            return add(component, Integer.valueOf(0))
+            addToLayer(component, componentLayer)
+            if (!interopBlending) {
+                bridge.addClipComponent(component)
+            }
+            return component
         }
 
         override fun remove(component: Component) {
-            bridge.clipComponents.remove(clipMap[component]!!)
-            clipMap.remove(component)
+            bridge.removeClipComponent(component)
             super.remove(component)
         }
 
+        private fun addToLayer(component: Component, layer: Int) {
+            if (renderApi == GraphicsApi.METAL) {
```

Comment:
I agree with Alexander, we use so many hacks for `transparency` that we can easily break the behavior in the future.

I tested main variations on Windows/macOs - it works, but to be safe, let's write `java.awt.Robot` screenshot tests for different OS'es, as we wrote them in Skiko (look at `SkiaLayerTest`).

Disable it for now, as we should run them only on CI. I will configure it later myself.

We can merge the PR now to receive feedback earlier, but If we agree that we can/should write the tests, we should write them before the 1.6.0 release (please, create a task).

Event if Robot tests can be flaky and it is hard to run them locally, they are very useful.

To write tests for transparency, we can place an opaque window behind a transparent window.

Reply:
https://youtrack.jetbrains.com/issue/COMPOSE-672